sycl/doc/OptionalDeviceFeatures.md

```diff
@@ -0,0 +1,707 @@
+goal is to ensure that two kernels or exported device functions are only
+bundled together into the same device image if they use exactly the same set
+of aspects.
+
+For the purposes of this analysis, the set of *Used* aspects is computed by
+taking the union of the aspects listed in the kernel's (or device function's)
+`!intel_used_aspects` and `!intel_declared_aspects` sets.  This is consistent
+with the SYCL specification, which says that a kernel decorated with
+`[[sycl::requires()]]` may only be submitted to a device that provides the
+listed aspects, regardless of whether the kernel actually uses those aspects.
+
+We must also split two kernels into different device images if they have
+different `[[sycl::reqd_sub_group_size()]]` or different
+`[[sycl::reqd_work_group_size()]]` values.  The reasoning is similar as the
+aspect case.  The JIT compiler currently raises an error if it tries to compile
+a kernel that has a required sub-group size if the size isn't supported by the
+target device.  The behavior for required work-group size is less clear.  The
+Intel implementation does not raise a JIT compilation error when compiling a
+kernel that uses an unsupported work-group size, but other backends might.
+Therefore, it seems safest to split device code based required work-group size
+also.
+
+Therefore, two kernels or exported device functions are only bundled together
+into the same device image if all of the following are true:
+
+* They share the same set of *Used* aspects,
+* They either both have no required work-group size or both have the same
+  required work-group size, and
+* They either both have the same numeric value for their required sub-group
+  size or neither has a numeric value for a required sub-group size.  (Note
+  that this implies that kernels decorated with
+  `[[intel::named_sub_group_size(automatic)]]` can be bundled together with
+  kernels that are decorated with `[[intel::named_sub_group_size(primary)]]`
+  and that either of these kernels could be bundled with a kernel that has no
+  required sub-group size.)
+
+These criteria are an additional filter applied to the device code split
+algorithm after taking into account the `-fsycl-device-code-split` command line
+option.  If the user requests `per_kernel` device code split, then each kernel
+is already in its own device image, so no further splitting is required.  If
+the user requests any other option, device code is first split according to
+that option, and then another split is performed to ensure that each device
+image contains only kernels or exported device functions that meet the criteria
+listed above.
+
+#### Create the "SYCL/image-requirements" property set
+
+The DPC++ runtime needs some way to know about the *Used* aspects, required
+sub-group size, and required work-group size of an image.  Therefore, the
+post-link tool provides this information in a new property set named
+"SYCL/image-requirements".
```

Comment:
this probably not the best place to discuss this, but having the word "image" used in identifiers for both device images and images (pictures) can be confusing. Can we call the former "device images", especially when used in property names and other identifiers?

Reply:
That's a good point.  I propose we rename the property set to "SYCL/device-requirements".  I think this makes sense because it lists the requirements that a device must have in order to use this binary image.

Any objections?

Reply:
The property set is renamed in 0e479767de1486f42f36c621cee5e2f3a18208fb.